Add clamped vector polylines

ebogo1 opened the pull request:
Opening as a draft to make sure things are in their place before adding specs.
- Moves `decodeVectorPolylinePositions` logic to its own file in Core
- Adds `examineVectorLinesFunction` and `minimumMaximumVectorHeights` properties to `Cesium3DTileset`
- Adds the `Vector3DTileClampedPolylines`class
- Adds logic for floating vs clamped polylines to `Vector3DTileContent`
- Adds vert and frag shaders for clamped polylines
- Adds a worker for creating clamped polylines

cesium-concierge commented:
Thanks for the pull request @ebogo1!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/clamp-vector-tiles/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/master/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

likangning93 commented on Source/WorkersES6/createVectorTileClampedPolylines.js:
```diff
@@ -0,0 +1,507 @@
+
+var previousCompressedCartographicScratch = new Cartographic();
+var currentCompressedCartographicScratch = new Cartographic();
+function removeDuplicates(uBuffer, vBuffer, heightBuffer, counts) {
```
This might be handy to have elsewhere in vector tiles too, wherever positions are getting unpacked from those quantized u/v buffers. It'll need a more specific name if it becomes its own file though.

likangning93 commented on Source/Shaders/Vector3DTileClampedPolylinesVS.glsl:
```diff
@@ -0,0 +1,86 @@
+
+void main()
+{
+    vec3 scratchNormal;
```
I think this originates from some paranoia that GLSL doesn't optimize how registers get used, but in practice it's probably fine to keep creating new `vec3` vars instead. I think that would be more readable.

ebogo1 replied:
@lilleyse any thoughts on this?

lilleyse replied:
I agree with @likangning93, should be fine to create a new `vec3` every time instead instead of a scratch.

ebogo1 commented:
I removed the unused property in `Cesium3DTileset`, marked the callback as `@experimental`, and renamed `CornerIds` + shader variables in a2ac850. Do you think any of the other changes are in scope for this PR?

ebogo1 commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -898,6 +898,13 @@ function Cesium3DTileset(options) {
+   *
+   * @private
+   */
+  this.examineVectorLinesFunction = undefined;
```
I think CI doesn't like using `@experimental` without also having an `options` object for it 🤔

lilleyse replied:
This seemed to pass CI in https://github.com/CesiumGS/cesium/pull/9398/commits/2736e093f133e3fd0b1b7002148987562a97879e

ebogo1 commented:
@likangning93 had some minor changes for CI to pass but should be ready now. I added the `examine` callback function as an option for the tileset constructor along the way.

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -885,7 +886,10 @@ function Cesium3DTileset(options) {
    * @default false
    */
-  this.debugShowMemoryUsage = defaultValue(options.debugShowMemoryUsage, false);
+  this.debugShowMtyemoryUsage = defaultValue(
```
```suggestion
  this.debugShowMemoryUsage = defaultValue(
```

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -898,6 +902,15 @@ function Cesium3DTileset(options) {
+   * Function for examining vector lines as they are being streamed.
+   *
+   * @type {Function}
+   * @default undefined
```
I don't think `@default undefined` is needed

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -898,6 +902,15 @@ function Cesium3DTileset(options) {
+   *
+   * @type {Function}
+   * @default undefined
+   * @experimental This feature is experimental and is subject to change without Cesium's standard deprecation policy.
```
CC https://github.com/CesiumGS/cesium/pull/9398#discussion_r585836031 I'm not sure if it's worth making this experimental since the callback signature isn't documented and one of the parameters is a `Cesium3DTileBatchTable` which isn't in the public API.

I would just keep this private and remove `examineVectorLinesFunction` from the constructor's doc as well.

lilleyse commented on Source/Shaders/Vector3DTileClampedPolylinesFS.glsl:
```diff
@@ -0,0 +1,52 @@
+    gl_FragColor = u_highlightColor;
+
+    czm_writeDepthClamp();
+}
```
Add newline

lilleyse commented:
@ebogo1 you mentioned it above, but make sure to address https://github.com/CesiumGS/cesium/pull/9399#discussion_r586419097 soon because otherwise polylines won't be clamped at all.

lilleyse commented:
Is there a sandcastle that we could play with? Screenshots? Maybe we can tile some non-customer data. @likangning93 do you know if we have any data like that?

Normally I would also suggest a new official sandcastle example but given that this is for the older `vctr` format I don't think we need one. However we should add a vector tile polylines (clamped) option to the [`Development/Picking`](http://localhost:8080/Apps/Sandcastle/index.html?src=development%2FPicking.html&label=Development) sandcastle to easily test the picking behavior.

Also, we should have units tests for the new behavior.

lilleyse commented:
The code is currently crashing because the `startFaceNormalAndVertexCorners` property rename was not applied to `createVectorTileClampedPolylines.js`. Just to reiterate a point, it's always a good idea to be testing new features with a sandcastle example of some sort to catch mistakes like this (unit tests being the other line of defense).

Here's a sandcastle I put together for testing: [Sandcastle](http://localhost:8083/Apps/Sandcastle/index.html#c=jVJRa9swEP4rwk8OGKlbyUvrhpWkg0LHCvHaF0GR5UuqTZbMSXZxR//7JCvpkrYPBWNJ39133+k7DQLJoOAJkFwQA09kCU71Lb2bsJxncjovrfFCGUCeFeQvN4R4QAzILdpBNYBne6JEEB7uLeqmSin5rODmZXbODTdDkJOiBRRBLulSJ8EATeiUxNhRZKttDdSjME73AZEjBSNqDU2o4bGH889QNhiu8F1IuNTdY1Q/ofPXlrzS4MC/7alD1SqvBnBUNE0er31gUVpOV1Ui55MthPSoz9KOEJ5RytK37kA6thJesCMiuwPpLe6WCN1aPergtWO7tuhvZ00wPlWVWjinNkoKr6ypxg5evV++C9Hl1fr614+H09VDdX1zFSu8zLhJw9iXd37UcDz+wxbXMZxuJ622YdQ8w20t8q/zeUGm30kR7Zzx7P+gk4/P1raVzXdKMZQVWTkJLvYmfVNtZ9FH4/JglIe20+EJOVb38k9oTzoXiTG1ZIfUslEDUc3FB480uRQim17rtXoGni1KFvLfUbUVjTLbnwOgFmNMe/yyuEkgpbRk4fgx01ura4FvKv8D)

![Selection_682](https://user-images.githubusercontent.com/915398/109898576-81b9aa80-7c62-11eb-948b-11a4dd3e29df.png)

What's interesting is that even after that problem is fixed there's still some issue where the polylines don't show up at all. Is it because of our usage of `ApproximateTerrainHeights` and the area being flat on the ocean?

lilleyse commented:
This one's kind of random, but in https://github.com/CesiumGS/cesium/pull/8726 I had to render shadow volumes with premultiplied alpha blending instead of regular alpha blending since alpha blending against a black framebuffer like the globe translucency framebuffer would result in darker colors than normal. The same sort of change needs to happen here - I pushed the change: https://github.com/CesiumGS/cesium/pull/9398/commits/0351120f0763b88677d20326cdcb31e2cc05e1fc.

This is what I was starting to test in the comment above. The commented out lines will enable globe translucency.

I think similar changes will be needed in https://github.com/CesiumGS/cesium/pull/9399 as well and I'll look at that once the changes here are confirmed to work.

lilleyse commented:
> What's interesting is that even after that problem is fixed there's still some issue where the polylines don't show up at all. Is it because of our usage of `ApproximateTerrainHeights` and the area being flat on the ocean?

False alarm, the sandcastle was wrong. `classificationType: Cesium.ClassificationType.CESIUM_3D_TILE` should have been `classificationType: Cesium.ClassificationType.TERRAIN`.

[Updated sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=jVJRa9swEP4rwk8OGKlbyUvrhpWkg0LHCvHaF0GR5UuqTZbMSXZxR//7JCvpkrYPBWNJ39133+k7DQLJoOAJkFwQA09kCU71Lb2bsJxncjovrfFCGUCeFeQvN4R4QAzILdpBNYBne6JEEB7uLeqmSin5rODmZXbODTdDkJOiBRRBLulSJ8EATeiUxNhRZKttDdSjME73AZEjBSNqDU2o4bGH889QNhiu8F1IuNTdY1Q/ofPXlrzS4MC/7alD1SqvBnBUNE0er31gUVpOV1Ui55MthPSoz9KOEJ5RytK37kA6thJesCMiuwPpLe6WCN1aPergtWO7tuhvZ00wPlWVWjinNkoKr6ypxg5evV++C9Hl1fr614+H09VDdX1zFSu8zLhJw9iXd37UcDz+wxbXMZxuJ622YdQ8w20t8q/zeUGm30kR7Zzx7P+gk4/P1raVzXdKMZQVWTkJLvYmfVNtZ9FH4/JglIe20+EJOVb38k9oTzoXiTG1ZIfUslEDUc3FB480uRQim17rtXoGni1KFvLfUbUVjTLbnwOgFmNMe/yyuEkgpbRk4fgx01ura4FvKv8D)

ebogo1 commented:
Can confirm clamping behavior is working on Cartersville TIN mesh after 4668ca6
![image](https://user-images.githubusercontent.com/31491650/110405427-9a550680-804e-11eb-9fb7-ac19a597a5b2.png)

lilleyse commented:
I realized I had to make some adjustments to the premultiplied alpha code because it wasn't taking into account the fact that the batch table overwrites `gl_FragColor`. I pushed the changes in https://github.com/CesiumGS/cesium/pull/9398/commits/7ada1a79c6529d17236b4211cacc84dab76ab583. This also makes b3dm classification tileset use premultiplied alpha which was missed in https://github.com/CesiumGS/cesium/pull/8726.

These are the sandcastles I used for testing:

[vctr-sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVTBbtswDP0VwScHCOR0Qy5NGqxLN2DAsBWN0V18UWQm0UZLhiS7SIf++ygpyZK0w7rDAMOyKD6S75FyLyzrFTyAZVdMwwObg1Ndw++jLa8yGfdzo71QGmyVDdnPSjPmwVqy3FrTqxrs5R4oLQgP34zFukwu+WBY6afBpNKV7imdFA1YQelSXu4kaODJenDyCsGBP/dqrWqUVz04Luo6D4UcFZ2WtzdlAuexUMY6i5fpi7Eq47xIz6IF6Yob4UVxAizuQXpjd0sw3RrcIrF3xa4s/t0ZTVKkqBKFc2qlpPDK6HLbwkGN+bMjXn64u7v+9CVAnwaVTrrs4zq/RTjtxHFti3CcaEmDhlSvMrteivzNeDxk8TUasgs+GlTZb82TgI/GNKXJd5nS0ULoWgrnEYKcpVmvEd533htNnS+t0A470t2zNZolhN6vBDqgpdMyMMrlBuQPqAdpKE56FTHcH6LILQctlgg1EdzhJq9BrSxN30ch4RrbTRicER9PDvRezaGP/XR/Z6FW7MzC2L83KE7Gn5tEHGKTgltkQgsDKuu/ZdyPxXHGqGI2zKYxy2wf5Z1qWmN9uDk53RQPTYt0q12x7EgVz6VzKQBj0+IYOq1Vz1R99cJ/I10TOll1iAv1SOM0mxbk/wyKRtRKr7/2YFFsg9vmYvY5GTnn04K2LyO9MbgU9izyLw), [b3dm-classification-sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVVRb9owEP4rpzwFCTl0rNW0QrUO+oDUadWg20ukyiQHeHN8yHbo6LT/vksMrKGwVaqKIqz4vrv77jufkyRwaQB/ymKpEWgGpVNmDhKm3bwAr7RDD54g09I5NVuDNOQXaHd2ZIBIzUpaWCm8Z0sfDN7DAJ0qC/G13ovTKKvfB2S8VAZtGrXhV2oAPFrLOzeWVipH+37rmFmUHr+R1fkkQOJWOzW/W+epSU3CtMGQLaRuMKkqWC7I09zKokBv14Ha1tzgFpbucBKMcc2ntHrHYUTmCzoqbYZiZqm4dAwb5fHbzruzs0dsQuHCZWhQLK0qlFcrdELmebzJXAE3sAeiYkKPDYHjRuJMekVmw+nWauhXtNJIiISfcd2oofQyadBPBsFbBffkpiFCsm3Ud0cmjc43AjaUKx3mjU5zm/e03GIbfFWD8HMVPlJsuwLsBV4vcdeRwROTGFyNR7ef7rrDu8no+mrblIPkhPNrPuVHKY4rcyCZkSY+jGlk51MZvzk9bUOnfk5Ep8WHd5PmX50/SCG0eyxNnknnNVbICc3nGj+W3pPhSZlYaZwuOaSHqXS4U51HZiZ5IHkpTVYFjbMFZj8wb4VZUjPY24HdiD6/9Op3oPy/fx1xyhIEaK0CL4BM7BVzBtmbOXfz92w5my15mbAvPWFHSg6n7L8Sv2r2o2JH7ahX57jYxvigiiVZXw11zNeTR76e+N52ybRktbzInAsBAHrJY9derlag8v6BL0Moji2zUuuxesA0uugljH/iqknm/Ln6vEKr5bqCLU4ursOmEKKX8OthT0+kp9LuRf4D), [geom-sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=zViNbts2EH4VzisQe3Ap2ZJ/kibFHDduAyRIELsthnkoaIm22cmkQVLOvMDvviMpOZbtpO5SYAsSSOTxjt99dzye4nmowxH9i8zmCUVqKu4ZnyCC5oJxjaJEpDHSLKGKmhFRio0ZjdFoCWveUzGjWi7zBXjIF0SiBaP3VKIzxOk96lLF0hn+ZOfKw1Jkx13BNWGcymGpih6GHCFNpYSZWykWLKbyJFeMJCWafhYyiQduSblSHfJV5c2QD7nn3VqcXYsTQIH+GKNbCuZQJ0kor6KuSNLZiBH0kbMFlYrpJboTI6FZpNAVGWHUjwjnxm0w0EnoV8JjKdBAiigSCUMwRNcAgqEuExGRCaOZpzkxBVfdI3g3cMKydS+VydqlS8HvqBKpjCgeSzHrKFh2GZdrzbBeWzvnWMQqopziuWQzpgG+wiSOy9m+OQePgbB7IiIpRHROQYkCz4n1QE8pGgs5IxoxhVQ6+kojjbRA0ZTwCUWMuyWpTqXxD8z2hEQzAcZiCtFKVBUpSk+sCH6mWs/ViedNmJ6mIxyJmeccfN/3gvi1xehpSam3iLT0DLSe3V95OV7HYp5WEdFM8MF3cWqeCA1LGHvw27dZ/I5o4hU0vG5hC5c0Nme8PHW/KsEhGY29LTzLOV2Hrrsjwt2L/uXH6y/Buy+Dy6uLQ6K3118Xy70irPQSzuaThPSN2FEC+Srg7Dw4WiLBY2YMqRP0u5tC8FJ69cDiFTo7O0NHUohx7cicQjiaRrl89LPvh71e76iKativDEt/VPeragEengutxWzbQhCcnzebh1kYiPkugOPjTueb6gZ7/QXYgy3dXq/dDoLDdMMd3U6nXt+vq2VKt1f3er6/s9o9V4/lrZBFEZlRSTAkhCmmLt4xVZpxmywnhfQgUsMb4UHZmQ5Dv9YKQ9wMw1rYCGphtmW93vAbNRzW/Fb7uFbzG/VMEDaOG2Hdx0ErbLXDMAjMdMUKhWSmrOxs+oGSGKroLdPR9E4kSbZ1AzZttoKg7Tf84LjVaGU7vPZxvdVs+q1aI2gHx+12MxM0cb1dAyztViP0fb/dyLdeF310yV0pAwiIjESqbe2KUikBWbJEUzaZJvCn4aoawwUCJc3Vmk3BmTsnmfwEpTymY7iU4tzLCVCbdN2R2iTXRtECyvHYKbgRM1toSZNE3CNANxVw5+CNUEJB5P05iejFArB+gNKcmHm4Afg81Z3I+FQep9y+gIVrkSp6DVYsnTN4MUV9yCsOvSFjvLHzPVFoLumCgVqRiKpxUFii1rPGAhujcuZZ5n95QwtnhiuVvKrsEWKb2EDopqxA4JsndUFrTbxdtTKcWsduWfQnuGaoXwcRIRPGOUho3FtbKJZbEJZzojDl8a1QtgxW3uT+/rTlcMHeo6uSwphvofqQ++Duy0dg+50rmLamNtMIQ4fFaXF/x2b1aTKdH3t0YLuC8d8urq5uPuN7uKE7yXxKyj5uGGVTYrKF/a2EtHfa9c3H/sWX65tPF+7E9SFLI6J0Qs0FNhCTSULPU6j8HDq6gSRcJSlQr9GIKJp3RabkjUmiKDyydC5HU2pAZwSbSBRn0LqVPPzOczedPaTDkpyMSPnVQ/fm6uZuhSX6BdUbjSpaz0x2Zkb5jOFmWHIWV47iFaKA/7+H5q6JIrR1OTw4OMXm4mVhemmjssHMevxEw/KdTYsL4+MF/O8alwOtPNe8fNvEsw3MgQiebGIO1H+ykXlC/5lmZlfj8X1VfeZg/X+Tabub+zFd8A/ohF/YDb+0I35pV/y9nfHzyWRLYalaOrUp8zZf9iubzYXU5juxDB+HmsLHIXzGK2+UQi3TOFKqkvUlp96m6mnMFojFZ3v+W+FyFSTjNEn67G+A//bUg/U7qomw3fANtH8JWZpl09rbKzeJMT71YLhfUwuRjIjcsvwP)

ebogo1 commented on Specs/Scene/Vector3DTileClampedPolylinesSpec.js:
```diff
@@ -147,7 +147,8 @@ describe(
         var features = [];
-        tileset.createFeatures(mockTileset, features);
+        var content = tileset._root._content;
+        content._polylines.createFeatures(mockTileset, features);
```
Is this the right idea to access `Vector3DTileClampedPolylines.createFeatures()`? Web debugger shows all the `content` fields are undefined.

ebogo1 commented:
Pushed some specs for the new code, based on `Vector3DTilePolylineSpec`. I used `xit` for now since the rendering is failing due to a bug I found: 
![image](https://user-images.githubusercontent.com/31491650/110822119-c050f580-825e-11eb-91fb-216ba3fe4a5c.png)
When viewing a clamped polyline tileset from above, artifacts like this begin to appear. I don't believe this is related to different LODs, and it happens both in Chrome and FF. This behavior doesn't occur with regular polylines with `clampToGround = true`.

likangning93 commented:
> When viewing a clamped polyline tileset from above, artifacts like this begin to appear. I don't believe this is related to different LODs, and it happens both in Chrome and FF. This behavior doesn't occur with regular polylines with clampToGround = true.

Here's the relevant chunks of `Vector3DTilesClampedPolylines`:

```js
import CullFace from "./CullFace.js";

//...

function getRenderState(mask3DTiles) {
  return RenderState.fromCache({
    cull: {
      enabled: true, // prevent double-draw. Geometry is "inverted" (reversed winding order) so we're drawing backfaces.
      face: CullFace.FRONT
    },
```

I don't know why this is the case though, since the geometry code in the vertex shader and the indices for each volume should basically be creating "inside-out" geometry already.
This also doesn't seem to be a complete fix, so there might still be something else at play. Maybe something's also depth-testing where it shouldn't.

[Here's the Sandcastle I'm testing with](http://localhost:8080/Apps/Sandcastle/index.html#c=vVbdc+I2EP9XNH4iUyLL2Bhz+ZjmknSGmd71JnC5F78IWxC1suRKMheuk/+9awubj3A0fQnDIHu/d3+7K1ZUoxVn35lGV0iy7+iWGV4V+LGh9VIva95vlbSUS6ZTr4/+SSVCvm+Z1kD7otWK50x/aFUzzahl35QW+cyJ9M76tQov6JLp9VZhx9+MC/aJllOmVzxjk33JXuMRoUqLzsu84iL/pPJKsK9aQKA3xjBr/Bl7tpVmxv9M4aTinmr7NJmknovhxR2a/V0xYx+YBPNghX1AVlcMeC9nF6lM5QrqktGCaQp1cQXCJmOSYUdthHx/jyPUkmpunwqe3bHSPn2sFoumrrXpi1fiS6HmDFtNpREVULI1ZpLOBcv/j8pCAzS/0YzdiPKpDpbgYZeBhapCVQ5TKDUvuOUrZjDN815dkB0o3BHezZzybvHdE0Kph7HvvtOSZca/o5b6e4r+I8us0pujJn1RYi2gh4y/CQv/aZSEhnJWM0GN4QueUcuVnK1L1mF9+4qFZ/cPDzeTzw7SVDrQWrvGrgXb7+fd2KY126WVKaGgEVNPL+e0NxgO+6j5IX0UYHKWel1DtLYL+syLqphClzM5LaHw91qrGuUBiRJMGmHXI3gh1jPV1i+HfuOyyQDttf4tdCg8URn24nA0HpABHsVhHCYDApGcB0EQ4iQJxgEJozgcRyHEFkR4kJA4DoejeBySIDnbVFFpzmBUN17Adc41IHDK6TnBY/dJouE4DMEPFAATQoJRFIMr8BGNyCAeJhBOwwCHhMRkGA+CIYQZBIPaf1We8rGjNhxHQTgYhqM+CnEQRQTyIGSURCOgsvNx7X7cfpJwlIyDNr+XzZnDbG+yAtMdSlMq84waK1jd2TOlxJzqj5W1SsKKWCpkLARVL7FFJV1ZemduoR3D7L1QO4IbuH4/5GC63wG7LXp7+IU1fvUYX7wVRthogNZxFJU0SjS7GCQ3ax42akk1LQxawKRmldZQ63a915sRboeLWrvemaUyvDF51bZES+lEtsh0Mh2pE4KKdtyqbNI6CE+6VZoCKz2d8+Pk/ts22SZXp3TQtI741r5NPfRLi0ebI37e0uoCH5VZv0Hmx64MdMA2tMNWd5z/7vZdN530qXi3Quu3CB2P+MRg7JqpylORAHd9kvvzar3svhxsPceoZ2b75DlJ18/NOHl977K5Ea9bF7/yolTa1ld6D65wy4pSwF8248+r7C+44zJjnD5Cl/6u6mXOV4jnV0f+Frr7GziLSogp/8FS7/rSB/lXqkLRnMvlHyumBV3XYk/B9e+OiDG+9OH1uKZ1Q3Fg+V8)

likangning93 commented:
There might also be something about the `CullFace.FRONT` thing combining with clamping the "min depth" of the polyline volumes to about `-1000` (the value it's getting from `ApproximateTerrainHeights` right now seems to be `-100000`) also producing better results, but... that's yet another mystery.

likangning93 commented:
## Investigating  "When viewing a clamped polyline tileset from above, artifacts like this begin to appear."

I made some shader and render state changes off 6e4d7ef to try to get to the bottom of these, resulting in the two commits below. Here's the [Sandcastle I used for testing.](http://localhost:8080/Apps/Sandcastle/index.html#c=nVZbc+I2FP4rKk9kSmQbg4HNZZrLZiYzm92dQLMvfhH2AdQIyZVksqST/95jCxuTUJItwyD7nO/cL2LFNFlxeAJNzoiEJ3IFhudL+lDS2nErKd+vlLSMS9Bxq0P+iSUhnmdBa6R912rFU9CfKtFEA7PwQ2mRThykfdQpRPiSzUGvtwINexMu4I5lY9ArnsDtLrJdWiQk16K2Ms25SO9Umgv4Uwt09MIYsMabwE+bazDeV4YnE5+Ztovb27jlfHhxh4a/czD2HiSqRy3wiVidA/Jejk5iGcsV5iVhS9AM8+ISRE0CEqijliDP2+EINWea28WSJ9eQ2cVlPpuVeS1Un7yBz4WaArWaSSNypCRrCpJNBaS/IjLTWJoblsCFyBaFsz7t1xFYzCpm5XUImeZLbvkKDGVp2i4S0iiFO8LriRNuJt89ERK3KPXcd5xBYrxrZpm3I+g9QGKV3hwF6bsSa4E9ZLyNW/QvoyQ2lNOaCGYMn/GEWa7kZJ1BXeurNyw6+Xx/f3H71ZU0lq5oYybThBkroIhropSYMn2ZW6skNohV87mAOieXD6Zo5lkuk0IraR+5xq6cS2Gaz8cL9XSpcplyOX9QIl8C5vK3dyAnhZqdhO+0W7tw9uV/eDzTAM9Avny7ft/xmxJ7o9kejxu8X3C1UmHsWsDusmgWflywXc8kSiic8ril51PW7vb7HVL++B0SUP8obtW661gse4Q7uGTJ40RNFqgMp7ed4hwX3E2cbgDpTKwnqmrOFF3mskSRnb1yhQrwicmwHYWDUdfv0kEURuGw66Mnx0EQhHQ4DEaBH/aicNQL0begR7tDP4rC/iAahX4wPNq0qNIccA9urKDplGtIDho99unIfYa9/igM0Q4mgPq+Hwx6EZpCG72B3436Q3SnZKBB34/8ftQN+uhmEHQL+3l2yEZDrD/qBWG3Hw46JKRBr+djHL4/GPYGSIXjUWF+VH2G4WA4Cqr4XjZnlXC0WD26QSu64f22nStiisLt79J9Fe5iOzQ6bR/Ed5B3beOcYIPsNZ0oaZQo9zQiN42O2zZjOAqGzBSu/FxrLHG1+ovJwJvDTQnu00wZXqo8q7qwotSQbUvUmJpUg7CYNTfPyrBeuSfdmo2RFR+O+eH2849tsGWsTujVnDjiR4clbpHfq21fxUh/bmlFgvdi1h/APDcx2Hxb115PmOO8P2ZNMzX6kL9b0PojoP0eH5jJppo8O+QJctcHuf+drZfmS2NocVY2jGKitk8th3T9XA5bq9M6LRf6eWXiD77MlLbFdd/G693CMhP4d8540zx5xPWfGOPkCTn1mqKnKV8Rnp7t+cvo7nbkzHIhxvwZ4tb5qYf4N6JCseIe/bYCLdi6gC2C8y+OSCk99fB1v6R1Q/FK878)

> fix clamped vector polylines not drawing when camera is inside volume by culling front faces da38f78

The algorithm we're using to draw clamped line segments does something like:
1. draw a "decal" that covers the whole onscreen area the line segment is likely to cover onscreen. Conceptually this is a little bit like a shadow volume, except it doesn't involve stencil buffers etc.
2. for each fragment in the decal, figure out if that fragment is part of the line segment

The goal is to draw the "inside" faces of the volume instead of "outside," so that when the camera enters the volume it should still see the decal instead of backface-culling it all.
The diagrams and `REFERENCE_INDICES` in `createVectorTileClampedPolylines.js` indicate that we try to use winding order to accomplish this "inside-drawing," but for some reason that doesn't seem to work. I simplified the vertex shader for clamped vector polylines to make the volumes a little easier to grok, and noticed that when we _crash through the surface_ of the volume, the volume (and therefor the clamped polyline) is no longer drawn:

![crashThroughTheSurface](https://user-images.githubusercontent.com/5395255/110964681-711fc900-8321-11eb-8b06-de7f48574353.gif)

In other words, we were drawing front/outside faces, not back/inside faces. Flipping the face culling (or reversing the indices in the geometry) fixes the problem, but I still don't fully understand why the geometry is drawing front/outside faces in the first place.
Here's a patch for further investigation:

[investigateFaceFlip6e4d7ef.patch.zip](https://github.com/CesiumGS/cesium/files/6131223/investigateFaceFlip6e4d7ef.patch.zip)

> use bounding volume that covers whole shadow volume for draw commands in Vector3DTileClampedPolylines 6602501

With face-flipping turned on and no other changes, "smaller" chunks of the polyline were still missing. Moving the camera along the line segment caused the artifacts to "slide" instead of staying _where they can't hurt us_.

![whereTheyCantHurtUs](https://user-images.githubusercontent.com/5395255/110965476-39fde780-8322-11eb-8839-a959b513beed.gif)

This looked a little like parallax to me, so I switched the debug bounding volumes back on. Lo and behold! Even drawing back/inside faces correctly (as evidenced by the "sides" of the current volume being visible), the bottom of the volume wasn't drawing, hence the lack of classification!

![farFromTheShallow](https://user-images.githubusercontent.com/5395255/110965892-aed12180-8322-11eb-8a72-3ec800ed364d.gif)

Now, this was drawing using the actual-size shadow volumes, which go something like -100,000 meters below and 9,000 meters above the ellipsoid in this area.
I noticed that the draw command bounding volumes though were using the 3D Tile's bounding sphere, _not_ its enormously deep and tall bounding volume. I think Cesium was using those "shallow" bounding spheres to compute the far plane for the top-down view. As a result the bottom part of the shadow volume was getting clipped by the depth plane, since in the top-down view it's _far from the shallow now._ Updating the draw command to use a bounding volume based on the shadow volume seems to fix that problem.

Side-note, I also noticed that `czm_depthClamp` (depthClamp.glsl) doesn't do anything when the renderer is in log-depth mode, which was also a helpful observation pointing at depth clipping being the problem.

ebogo1 commented:
With some more help from @likangning93 I was able to get rendering and picking specs to pass as well. I think this is good for another round of review!

lilleyse commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -1654,6 +1661,20 @@ Object.defineProperties(Cesium3DTileset.prototype, {
+    get: function () {
+      return this._minimumMaximumVectorHeights;
+    },
+  },
```
Bump - I think this can be removed?

lilleyse commented:
I renamed `noClassificationModels` to `vectorClassificationOnly` and marked it as `@experimental` since other vctr-related properties are also marked as experimental.

Make sure to do the same rename in SC once the newer version of CesiumJS is pulled in.

lilleyse commented:
I also left a comment above: https://github.com/CesiumGS/cesium/pull/9398#pullrequestreview-602167702

lilleyse commented:
By the way, excellent detective work @likangning93

lilleyse commented:
Thanks @ebogo1 @likangning93 !
